server/src/main/java/org/elasticsearch/action/ActionModule.java

```diff
@@ -448,12 +445,9 @@ public ActionModule(Settings settings, IndexNameExpressionResolver indexNameExpr
                     throw new IllegalArgumentException("Cannot have more than one plugin implementing a REST wrapper");
                 }
                 restWrapper = newRestWrapper;
-                if (restWrapper.getClass().getCanonicalName() == null ||
-                    restWrapper.getClass().getCanonicalName().startsWith("org.elasticsearch") == false) {
-                    deprecationLogger.deprecate(DeprecationCategory.PLUGINS, "3rd_party_rest_deprecation", "The " +
-                        plugin.getClass().getName() + "plugin installs a custom REST wrapper. This functionality is deprecated and will " +
-                        "not be possible in Elasticsearch 8.0. If this plugin is intended to provide security features for Elasticsearch " +
-                        "then you should switch to using the built-in Elasticsearch features instead.");
+                if ("org.elasticsearch.xpack.security.Security".equals(plugin.getClass().getCanonicalName()) == false) {
+                    logger.warn("The " + plugin.getClass().getName() + " plugin tried to install a custom REST wrapper. This " +
+                        "functionality is not available anymore, and the plugin has been disabled.");
                 }
```

Comment:
The code here only logs a warning when a custom plugin tries to install its own rest warpper. It doesn't really try to do anything to stop it though the message says "the plugin has been disabled"? From what I can tell, it would still go ahead and install the plugin and its rest wrapper?

Reply:
It won't. Because of the change in `Security::getRestHandlerWrapper` ^^, xpack security rest handler will always be installed even when security is disabled. But you have a point here, this message will never be reached because we throw an exception two lines above if a second plugin is trying to install a rest handler. There are several options on how we can handle it:
1) Log the warning and throw (current implementation)
2) Log the warning, but only throw if it was an `es` plugin. Proceed ignoring 2nd rest handler if it was a 3rd party plugin
3) Log the warning and proceed ignoring 2nd rest handler regardless of the source of the plugin

I don't have strong preferences choosing between the options.  @bytebilly, what do you think?

Reply:
There is no guarantee that the Security plugin is the first ActionPlugin that installs a RestWrapper. We don't have any logic to ensure that. It is entirely possible that a custom plugin comes here first and installs the rest wrapper and in turn blocks the actual Security plugin.

When the above happens, the message will be logged. But it is misleading because nothing is disabled. The overall behaviour is more or less correct because the node will fail to launch (due to Security plugin trying to install the rest wrapper again). But I think the logic is confusing. I'd suggest: 
1. Throw an exception here and possibly remove the above exception throwing. 
2. Reword the message to make it clear a custom rest wrapper triggers an exception

Reply:
Could we check if the REST wrapper is being installed by the x-pack plugin, and throw a warning+exception otherwise? This should happen for each REST wrapper, including the first one.

If I understand them correctly, other proposals seem not hitting the target:

> Log the warning and throw (current implementation)
This doesn't ensure that the REST wrapper is from the x-pack plugin

> Log the warning, but only throw if it was an es plugin. Proceed ignoring 2nd rest handler if it was a 3rd party plugin
Not sure to get it, but we should avoid any kind of third-party plugin at all

> Log the warning and proceed ignoring 2nd rest handler regardless of the source of the plugin
This doesn't ensure that the REST wrapper is from the x-pack plugin

Reply:
I actually though we do have the logic to ensure `es` plugind installed firs through plugin dependancy sorting... But now I see you are right.
Changing the logic to throw on any non `es` security plugin trying to install the handler.